src/decoder.rs

```diff
@@ -14,8 +14,10 @@ use crate::systematic_constants::num_ldpc_symbols;
 use crate::systematic_constants::{
     calculate_p1, extended_source_block_symbols, num_lt_symbols, num_pi_symbols, systematic_index,
 };
+use serde::{Deserialize, Serialize};
 use std::collections::HashSet;
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
```

Comment:
What's the use case for Clone, Serialize, Deserialize, PartialEq, and Eq? I don't think Decoder or Encoder, or their related classes need to support those. It's possible we'll want to add something like a thread local cache in the future, and I'd rather not restrict the internals of these classes by promising to support serialize...etc on them

Reply:
See my reply for the comment above.  Serialize and Deserialize support checkpointing, PartialEq and Eq are for testing (verifying that Serialize and Deserialize worked as expected), and because the state information I'm creating need them, as well as Clone.  Basically, all of this is due to the aforementioned orphan rules.  See the [API guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits) for more information.

Reply:
Thanks, I understand your use case now!